### ISSUE TITLE
Packaging ROS for homebrew

### ISSUE THREAD

**Issue by NikolausDemmel** [ros-infrastructure/bloom#254]
I'm opening this ticket to collect the thoughts and ideas that people have already had about this.

@wjwwood: Could you please copy your answer from [1] over here as a comment. I don't have enough karma to edit your post to get at the source.

I would say that generating formulae automatically with bloom is the only viable solution (as opposed to manually maintaining formulae). However, I actually started bottom up and experimented a little by manually creating some formulae (for now catkin, rospack, roslib) to see how the generated formulae should look like [2].

This has already revealed some issues. Most notably homebrew does at the moment not allow to specify a custom path where to install the formula contents. We would want to install e.g. in `/usr/local/ros/hydro` rather than `/usr/local/` (or ideally `/opt/ros/hydro`, but homebrew has a strong promise to only touch /usr/local). Moreover, currently homebrew installs only stuff in the default FHS folders (bin, lib, etc, ...), but files/folders on the toplevel do not get installed (setup.bash, ...). Another issue is the question about python dependecies. On debian they all have apt-equivalents, but in brew world that is not done. However, there is the possibility for us to maintain such formulae for all the python module dependecies at least for the core packages. On the other hand, homebrew can at least depend on pip installed modules, such that their precense is checked. But it won't automatically call pip install for you. Maybe something in conjunction with rosdep can be done here. I would prefer sticking with pip for installing these.

Anyway, the install prefix issue has to be sorted out before automatic generation can be tackled. I asked the homebrew mailinglist, but didn't quite get enough info so far [3].

[1] http://answers.ros.org/question/130346/packaging-ros-with-homebrew/
[2] https://github.com/NikolausDemmel/homebrew-ros_experimental
[3] http://librelist.com/browser//homebrew/2014/3/9/installing-group-of-formulae-into-a-different-folder/

**Comment by NikolausDemmel** [comment]
After having to reinstall on my mac from source once again, I had a very brief look at the state of this. The generator proof of concept is out of date with its dependencies (e.g. rosdep) and doesn't work as is. The easiest would maybe be to start again with the current debian / rpm generators and adapt them into homebrew geneators with the ideas from the proof of concept code.

Concerning the critical issues mentioned above (https://github.com/ros-infrastructure/bloom/issues/254#issuecomment-37401069), I had some thoughts for workarounds. I'll just dump it here in case some day someone picks this up:

Homebrew doesn't allow to link kegs into a custom prefix like `/opt/ros/lunar/`, i.e. only the default prefix (`/usr/local`), which means the linking the formulae for the same pacakge in multiple ROS distros would conflict. Moreover, homebrew doesn't allow to link files in the top-level (directly into `/usr/local/`), but instead only in the standard FHS subfolders like lib,bin,share, etc. This means that we cannot link into /usr/local/ and have homebrew create the `setup.sh` files when linking the catkin keg.

The linking code in homebrew is part of the `keg` class, and thus it seems it cannot be changed by writing some custom code in your formulae. In homebrew we will likely meet some resistance to add functionality to work around this limitation upstream.

A workaround could be to link everything into some custom folder, inside one of the allowed subfolders, maybe `/usr/local/share/opt/ros/lunar`. This would allow to have multiple distros installed, and requires no changes to homebrew. It would be great if that would then be usable from a standard path by having a symlink `/opt/ros/` --> `/usr/local/share/opt/ros/`.

Additionally, formulae have a post-install script, which is called when a formula is installed (both from source and also from a bottle). We could have some custom code e.g. to setup the `/opt/ros/` --> `/usr/local/share/opt/ros/` link if it doesn't exist.

Alternatively, we could make all formulae keg-only and add an external command to `brew` (can be setup by just added our Tap) which does the linking of the kegs directly into /opt/ros/..., e.g. with `brew ros link ....`. We could even (ab-)use the post-install script to call this automatically. The disadvantage is that it is more complicated to implement with all corner cases, and also, it isn't such a great user-experience, since our formulae would behave even more differently from regular homebrew formulae.

One could also use `brew ros` not just for linking the kegs into `/opt/ros`, but as a general interface for installing packages, i.e. make `brew ros install ...` install the formula and do the linking. While this again makes things even more complex to implement, it gives even better control and would even allow to do things like installing pip dependencies (maybe using rosdep), which homebrew itself doesn't support. However, it might be strange if `brew install foo-pkg` is possible, but doesn't do the right thing. 

The cost of maintaining this with ever-changing homebrew might be quite high, so I would favour a solution with little as possible dependence on homebrew internals.

As for the dependency resolution speed issue, I think it is still too slow. However, in the linked upstream issue, one of the homebrew contributors mentioned he had prototyped and tested a possible solution, but it seems like that never got anywhere (also because noone was pushing for it). If we raise the need again, we might get some help to get this fixed in homebrew.

**Comment by wjwwood** [comment]
I agree with the sentiment that macOS progress is gated on the ability to regularly build packages and to have enough manpower to overcome the volatility in Homebrew. Until you can have a job that just builds all of desktop-full (for example) regularly in a CI fashion and enough people to keep it up to date, then resolving packaging issues doesn't make a lot of sense to me.

It's worth pointing out though, that having the ability to bottle what works might make travis more viable because you don't have to build everything in one job, instead you can pour bottles for dependencies and build only what's new. Even then there will be packages that don't work, but at least it might push the problem down the line a bit.

My efforts on packaging (back when I had time to spend on this topic) were forestalled by not having reproducible builds as well. And things like brew bot and travis CI didn't exist.
